File: Maximum Score From Removing Substrings/Maximum Score From Removing Substrings.py
# Solution for Maximum Score From Removing Substrings in PY

class Solution:
    def removeSubstr(self, s: str, pattern: str) -> str:
        stack = []
        for ch in s:
            stack.append(ch)
            if len(stack) >= 2 and stack[-2] == pattern[0] and stack[-1] == pattern[1]:
                stack.pop()
                stack.pop()
        return ''.join(stack)

    def maximumGain(self, s: str, x: int, y: int) -> int:
        first, second = ("ab", "ba") if x > y else ("ba", "ab")
        afterFirst = self.removeSubstr(s, first)
        firstGain = (len(s) - len(afterFirst)) // 2 * max(x, y)
        afterSecond = self.removeSubstr(afterFirst, second)
        secondGain = (len(afterFirst) - len(afterSecond)) // 2 * min(x, y)
        return firstGain + secondGain

File: Maximum Score From Removing Substrings/Maximum Score From Removing Substrings.java
// Solution for Maximum Score From Removing Substrings in JAVA

class Solution {
    private String removeSubstr(String s, String pattern) {
        StringBuilder sb = new StringBuilder();
        for (char ch : s.toCharArray()) {
            sb.append(ch);
            int n = sb.length();
            if (n >= 2 && sb.charAt(n - 2) == pattern.charAt(0) && sb.charAt(n - 1) == pattern.charAt(1)) {
                sb.setLength(n - 2);
            }
        }
        return sb.toString();
    }

    public int maximumGain(String s, int x, int y) {
        String first = x > y ? "ab" : "ba";
        String second = x > y ? "ba" : "ab";

        String afterFirst = removeSubstr(s, first);
        int firstGain = (s.length() - afterFirst.length()) / 2 * Math.max(x, y);

        String afterSecond = removeSubstr(afterFirst, second);
        int secondGain = (afterFirst.length() - afterSecond.length()) / 2 * Math.min(x, y);

        return firstGain + secondGain;
    }
}
